Modules/TPC/src/TrendingTaskTPC.cxx

```diff
@@ -0,0 +1,292 @@
+      if (auto qo = qcdb.retrieveQO(dataSource.path + "/" + dataSource.name, timestamp)) {
+        mReductors[dataSource.name]->update(qo.get(), mSources[dataSource.name],
+                                            dataSource.axisDivision);
+      }
+    } else {
+      ILOG(Error, Support) << "Data source '" << dataSource.type << "' unknown." << ENDM;
+    }
+  }
+
+  mTrend->Fill();
+  ILOG(Info, Support) << "'mTrend' has been filled with the current entry." << ENDM;
+}
+
+void TrendingTaskTPC::generatePlots()
+{
+  if (mTrend->GetEntries() < 1) {
+    ILOG(Info, Support) << "No entries in the trend so far, no plot generated." << ENDM;
+    return;
+  }
+
+  ILOG(Info, Support) << "Generating " << mConfig.plots.size() << " plots." << ENDM;
+  for (const auto& plot : mConfig.plots) {
+    // Delete the existing plots before regenerating them.
+    if (mPlots.count(plot.name)) {
+      getObjectsManager()->stopPublishing(plot.name);
+      delete mPlots[plot.name];
+    }
+
+    // Draw the trending on a new canvas.
+    TCanvas* c = new TCanvas();
+    c->SetName(plot.name.c_str());
+    c->SetTitle(plot.title.c_str());
+    drawCanvas(c, plot.varexp, plot.selection, plot.option, plot.graphErrors, plot.name);
+
+    // Postprocess each pad (titles, axes, flushing buffers).
+    for (int p = 1; p <= mNumberPads; p++) {
+      c->cd(p);
+      if (auto histo = dynamic_cast<TGraph*>(c->cd(p)->GetPrimitive("Graph"))) {
+        /*
+                // Set a centered, nicely formatted title.
+                ///if (auto title = static_cast<TPaveText*>(c->GetPrimitive("title"))) {
+                if (auto title = dynamic_cast<TPaveText*>(histo->GetTitle())) {
+                  title->SetBBoxCenterX(c->GetBBoxCenter().fX);
+                  title->Draw();
+                } else {
+                  ILOG(Error, Devel) << "Could not get the title 'TPaveText' of the plot '"
+                    << plot.name << "'." << ENDM;
+                }*/
+        /// LOKI
+
+        // Set the user-defined range on the y axis if needed.
+        if (!plot.graphYRange.empty()) {
+          std::size_t posDivider = plot.graphYRange.find(":");
+          std::string yMinString(plot.graphYRange.substr(0, posDivider));
+          std::string yMaxString(plot.graphYRange.substr(posDivider + 1));
+
+          float yMin = std::stof(yMinString);
+          float yMax = std::stof(yMaxString);
+          histo->SetMinimum(yMin);
+          histo->SetMaximum(yMax);
+        }
+
+        // Configure the time for the x axis.
+        if (plot.varexp.find(":time") != std::string::npos) {
+          histo->GetXaxis()->SetTimeDisplay(1);
+          histo->GetXaxis()->SetNdivisions(505);
+          histo->GetXaxis()->SetTimeOffset(0.0);
+          histo->GetXaxis()->SetTimeFormat("%Y-%m-%d %H:%M");
+        }
+
+        // Manually empty the buffers before visualising the plot.
+        // histo->BufferEmpty(); // TBD: Should we keep it or not?
```

Comment:
Will this buffer emptying be needed? Should we re-enable it? Or can this be removed?

Reply:
We tried to re-enable it and the build crashes because "TGraph" class does not have the BufferEmpty method.